sdk/servicebus/service-bus/test/perf/azure-sb-package/send.ts

```diff
@@ -0,0 +1,125 @@
+/*
+# Overview
+Measures the maximum throughput of `sender.send()` in package `azure-sb`.
+
+# Instructions
+1. Create a Service Bus namespace with `Tier=Premium` and `Messaging Units=4`.  It is recommended to use the largest possible namespace to allow maximum client throughput.
+2. Create a queue inside the namespace.
+3. Set env vars `SERVICE_BUS_CONNECTION_STRING` and `SERVICE_BUS_QUEUE_NAME`.
+4. `ts-node send.ts [maxInflightMessages] [totalMessages]`
+5. Example: `ts-node send.ts 1000 1000000`
+ */
+
+import { createServiceBusService, ServiceBusService } from "azure-sb";
+import delay from "delay";
+import moment from "moment";
+
+const _payload = Buffer.alloc(1024);
+const _start = moment();
+
+let _sent = 0;
+let _accepted = 0;
+
+async function main(): Promise<void> {
+  // Endpoint=sb://<your-namespace>.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=<shared-access-key>
+  const connectionString = process.env.SERVICE_BUS_CONNECTION_STRING as string;
+  const entityPath = process.env.SERVICE_BUS_QUEUE_NAME as string;
+
+  const maxInflight = process.argv.length > 2 ? parseInt(process.argv[2]) : 1;
+  const messages = process.argv.length > 3 ? parseInt(process.argv[3]) : 10;
+  log(`Maximum inflight messages: ${maxInflight}`);
+  log(`Total messages: ${messages}`);
+
+  const writeResultsPromise = WriteResults(messages);
+
+  await RunTest(connectionString, entityPath, maxInflight, messages);
+
+  await writeResultsPromise;
+}
+
+async function RunTest(
+  connectionString: string,
+  entityPath: string,
+  maxInflight: number,
+  messages: number
+): Promise<void> {
+  const sbService: ServiceBusService = createServiceBusService(connectionString);
+
+  function sendMessages(): void {
+    while (_sent < messages && inflight() < maxInflight) {
+      _sent++;
+      sbService.sendQueueMessage(entityPath, { body: JSON.stringify(_payload) }, function(
+        err: any
+      ) {
+        if (err) {
+          console.log(err.message);
+        } else {
+          ++_accepted;
+          sendMessages();
```

Comment:
Just wondering... I might be completely off track here.
If I was using a callback way of sending messages, would I really wait for a message to be sent successfully, before sending the next message?
I would probably fire a lot of sends, keep track of the sends that failed, and then re-send the failed ones.

Reply:
I think there are two independent concerns:

1. How are messages sent to maximize throughput (up to the `maxInflightMessages` setting)?
2. How are errors handled?

I believe this test is correctly sending messages to maximize throughput.  Note that it does **not** wait until a single message is accepted before sending the next message.  The `sendMessages()` function will send messages in a loop until the inflight queue is full.  It's called once at the start of the test, and once again every time a message is received.  It's true that after the initial batch of sends, it's likely (or perhaps even guaranteed since JS is single-threaded) the loop will only be sending a single message at a time (since one accept means there's one slot open in the inflight queue).  But I think this is actually the best way to maximize throughput.

Error handling is more subtle.  It's easiest to say that perf results are only valid when no errors happened.  If the errors are infrequent enough, when an error happens just abandon the perf run and manually retry.  I believe this should be sufficient for Service Bus.

However, say errors happen often enough that every perf run has at least one error, and it's expected that developers handle them properly (by say retrying).  In this case, we should measure the overall performance (including retries) in the presence of errors.  But this adds a potential problem, where say one run had no errors and was faster, while another run had many errors and was slower.  So now you need to either track the number of errors and attempt to compensate  the perf results, or you need to ensure all runs have exactly the same number of errors.